File: plugins/rummy-plugin/Rummy.Game/DrawMove.cs
namespace Rummy
{
	public class DrawMove : Move
	{
		private readonly PileName _pile;

		public DrawMove(Player player, PileName pile)
			: base(player)
		{
			_pile = pile;
		}

		public override bool Equals(object obj)
		{
			if (obj == null || obj.GetType() != typeof(DrawMove))
				return false;

			var theOther = (DrawMove)obj;

			if (theOther.Player != this.Player)
				return false;

			if (theOther._pile != this._pile)
				return false;

			return true;
		}

		public override int GetHashCode()
		{
			return (int)Player + (int)_pile * 2;
		}

		public override void Realize(GameState gameState)
		{
			gameState.DrawCard(Player, _pile);
		}

		public override string ToString()
		{
			return string.Format("Player {0} -- Draw from {1} pile", Player, _pile);
		}
	}
}

File: plugins/rummy-plugin/srummy/java/edu/wpi/always/srummy/game/DrawMove.java
package edu.wpi.always.srummy.game;

/**
 * A class for Rummy game draw move.
 * @author Morteza Behrooz
 * @version 1.0
 */
public class DrawMove extends Move{

	private final Pile pile;
	
	public DrawMove(Player thePlayer, Pile aPile){
		
		super(thePlayer);
		pile = aPile;
		
	}
	
	public Pile getPile(){
		
		return pile;
		
	}

	/**
	 * Checks if another DrawMove is semantically 
	 * the same one by Player and Pile criteria.
	 * @param Object someMove
	 * @return true if same Player and Pile (semantically same Draw).
	 * @see TestDrawMove
	 */
	@Override
	public boolean equals(Object someMove){
		
		if(someMove == null 
				|| !(someMove instanceof DrawMove))
			return false;
		
		DrawMove supposedlySameDrawMove 
				= (DrawMove) someMove;
		
		if(!(supposedlySameDrawMove.getPlayer()
				== this.player))
			return false;
		
		if(supposedlySameDrawMove.getPile()
				!= this.pile)
			return false;
		
		return true;
	
	}
	
	/**
	 * Overides hashCode() of Object as equals() is overriden
	 * (In Java, hashCode() must return the same 
	 * value for two objects which are equal)
	 * @return int hashCode
	 * @see DrawMove
	 */
	@Override
	public int hashCode(){
		
		return player.ordinal()
				+ 2 * pile.ordinal()
				;
		
	}
	
	/**
	 * States the discard move as String. 
	 * @return String of discard statement
	 */
	@Override
	public String toString(){
		
		return 
				player.name()
				+ " just drew a card from "
				+ pile.toString()
		;
		
	}
	
	@Override
	public void happen(GameState gameState) {

		gameState.drawCard(player, pile);
		
	}
	
}
